// @mastra/server/dist/chunk-424T5F2J.cjs
'use strict';

var chunkPZQDCRPV_cjs = require('./chunk-PZQDCRPV.cjs');
var chunk75ZPJI57_cjs = require('./chunk-75ZPJI57.cjs');

// src/server/handlers/scores.ts
var scores_exports = {};
chunk75ZPJI57_cjs.__export(scores_exports, {
  getScorerHandler: () => getScorerHandler,
  getScorersHandler: () => getScorersHandler,
  getScoresByEntityIdHandler: () => getScoresByEntityIdHandler,
  getScoresByRunIdHandler: () => getScoresByRunIdHandler,
  getScoresByScorerIdHandler: () => getScoresByScorerIdHandler,
  saveScoreHandler: () => saveScoreHandler
});
async function getScorersFromSystem({
  mastra,
  runtimeContext
}) {
  const agents = mastra.getAgents();
  const workflows = mastra.getWorkflows();
  const scorersMap = /* @__PURE__ */ new Map();
  for (const [agentId, agent] of Object.entries(agents)) {
    const scorers = await agent.getScorers({
      runtimeContext
    }) || {};
    if (Object.keys(scorers).length > 0) {
      for (const [scorerId, scorer] of Object.entries(scorers)) {
        if (scorersMap.has(scorerId)) {
          scorersMap.get(scorerId)?.agentIds.push(agentId);
        } else {
          scorersMap.set(scorerId, {
            workflowIds: [],
            ...scorer,
            agentIds: [agent.name]
          });
        }
      }
    }
  }
  for (const [workflowId, workflow] of Object.entries(workflows)) {
    const scorers = await workflow.getScorers({
      runtimeContext
    }) || {};
    if (Object.keys(scorers).length > 0) {
      for (const [scorerId, scorer] of Object.entries(scorers)) {
        if (scorersMap.has(scorerId)) {
          scorersMap.get(scorerId)?.workflowIds.push(workflowId);
        } else {
          scorersMap.set(scorerId, {
            agentIds: [],
            ...scorer,
            workflowIds: [workflowId]
          });
        }
      }
    }
  }
  return Object.fromEntries(scorersMap.entries());
}
async function getScorersHandler({ mastra, runtimeContext }) {
  const scorers = await getScorersFromSystem({
    mastra,
    runtimeContext
  });
  return scorers;
}
async function getScorerHandler({
  mastra,
  scorerId,
  runtimeContext
}) {
  const scorers = await getScorersFromSystem({
    mastra,
    runtimeContext
  });
  const scorer = scorers[scorerId];
  if (!scorer) {
    return null;
  }
  return scorer;
}
async function getScoresByRunIdHandler({
  mastra,
  runId,
  pagination
}) {
  try {
    const scores = await mastra.getStorage()?.getScoresByRunId?.({
      runId,
      pagination
    }) || [];
    return scores;
  } catch (error) {
    return chunkPZQDCRPV_cjs.handleError(error, "Error getting scores by run id");
  }
}
async function getScoresByScorerIdHandler({
  mastra,
  scorerId,
  pagination,
  entityId,
  entityType
}) {
  try {
    const scores = await mastra.getStorage()?.getScoresByScorerId?.({
      scorerId,
      pagination,
      entityId,
      entityType
    }) || [];
    return scores;
  } catch (error) {
    return chunkPZQDCRPV_cjs.handleError(error, "Error getting scores by scorer id");
  }
}
async function getScoresByEntityIdHandler({
  mastra,
  entityId,
  entityType,
  pagination
}) {
  try {
    let entityIdToUse = entityId;
    if (entityType === "AGENT") {
      const agent = mastra.getAgentById(entityId);
      entityIdToUse = agent.id;
    } else if (entityType === "WORKFLOW") {
      const workflow = mastra.getWorkflowById(entityId);
      entityIdToUse = workflow.id;
    }
    const scores = await mastra.getStorage()?.getScoresByEntityId?.({
      entityId: entityIdToUse,
      entityType,
      pagination
    }) || [];
    return scores;
  } catch (error) {
    return chunkPZQDCRPV_cjs.handleError(error, "Error getting scores by entity id");
  }
}
async function saveScoreHandler({ mastra, score }) {
  try {
    const scores = await mastra.getStorage()?.saveScore?.(score) || [];
    return scores;
  } catch (error) {
    return chunkPZQDCRPV_cjs.handleError(error, "Error saving score");
  }
}

exports.getScorerHandler = getScorerHandler;
exports.getScorersHandler = getScorersHandler;
exports.getScoresByEntityIdHandler = getScoresByEntityIdHandler;
exports.getScoresByRunIdHandler = getScoresByRunIdHandler;
exports.getScoresByScorerIdHandler = getScoresByScorerIdHandler;
exports.saveScoreHandler = saveScoreHandler;
exports.scores_exports = scores_exports;
